ossdbtoolsservice/language/completion/mysqlcompleter.py

```diff
@@ -0,0 +1,950 @@
+            left, right = (child, par) if left == child else (par, child)
+            for rtbl in coldict[right]:
+                add_cond(left.col, right.col, rtbl.ref, 2000, 'fk join')
+        # For name matching, use a {(colname, coltype): TableReference} dict
+        coltyp = namedtuple('coltyp', 'name datatype')
+        col_table = list_dict((coltyp(c.name, c.datatype), t) for t, c in cols)
+        # Find all name-match join conditions
+        for c in (coltyp(c.name, c.datatype) for c in lcols):
+            for rtbl in (t for t in col_table[c] if t.ref != ltbl.ref):
+                prio = 1000 if c.datatype in (
+                    'integer', 'bigint', 'smallint') else 0
+                add_cond(c.name, c.name, rtbl.ref, prio, 'name join')
+
+        return self.find_matches(word_before_cursor, conds, meta='join')
+
+    def get_function_matches(self, suggestion, word_before_cursor, alias=False):
+        if suggestion.usage == 'from':
+            # Only suggest functions allowed in FROM clause
+            def filt(f):
+                return not f.is_aggregate and not f.is_window
+        else:
+            alias = False
+
+            def filt(f):
+                return True
+        arg_mode = {
+            'signature': 'signature',
+            'special': None,
+        }.get(suggestion.usage, 'call')
+        # Function overloading means we way have multiple functions of the same
+        # name at this point, so keep unique names only
+        funcs = set(
+            self._make_cand(f, alias, suggestion, arg_mode)
+            for f in self.populate_functions(suggestion.schema, filt)
+        )
+
+        matches = self.find_matches(word_before_cursor, funcs, meta='function')
+
+        if not suggestion.schema and not suggestion.usage:
+            # also suggest hardcoded functions using startswith matching
+            predefined_funcs = self.find_matches(
+                word_before_cursor, self.functions, mode='strict',
+                meta='function')
+            matches.extend(predefined_funcs)
+
+        return matches
+
+    def get_schema_matches(self, suggestion, word_before_cursor):
+        schema_names = self.dbmetadata['tables'].keys()
+
+        if suggestion.quoted:
+            schema_names = [self.escape_schema(s) for s in schema_names]
+
+        return self.find_matches(word_before_cursor, schema_names, meta='schema')
+
+    def get_from_clause_item_matches(self, suggestion, word_before_cursor):
+        alias = self.generate_aliases
+        s = suggestion
+        t_sug = Table(s.schema, s.table_refs, s.local_tables)
+        v_sug = View(s.schema, s.table_refs)
+        f_sug = Function(s.schema, s.table_refs, usage='from')
+        return (
+            self.get_table_matches(t_sug, word_before_cursor, alias)
+            + self.get_view_matches(v_sug, word_before_cursor, alias)
+            + self.get_function_matches(f_sug, word_before_cursor, alias)
+        )
```

Comment:
Are we using any of these get_from_clause_item_matches method? or other get_ methods above? I don't know what they are doing so should not be part of this PR. We shouldn't add anything extra that we do not understand or need for mysql literals.

Reply:
I think you can get rid of 80% of this file. We will only add stuff when needed.